utils/tfw_logger.cc

```diff
@@ -0,0 +1,463 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include <fcntl.h>
+#include <signal.h>
+#include <spdlog/sinks/basic_file_sink.h>
+#include <spdlog/spdlog.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include <fstream>
+#include <future>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+
+#include <boost/program_options.hpp>
+#include <clickhouse/base/socket.h>
+#include <clickhouse/client.h>
+
+#include "../fw/access_log.h"
+#include "clickhouse.hh"
+#include "mmap_buffer.hh"
+#include "error.hh"
+
+namespace po = boost::program_options;
+
+#define FILE_PATH	"/dev/tempesta_mmap_log"
+#define TABLE_NAME	"access_log"
+
+constexpr char pid_file_path[] = "/var/run/tfw_logger.pid";
+constexpr size_t WAIT_FOR_FILE = 1;  /* s */
+constexpr size_t WAIT_FOR_STOP = 10; /* ms */
+
+typedef struct {
+	const char		*name;
+	clickhouse::Type::Code	code;
+} TfwField;
+
+static const TfwField tfw_fields[] = {
+	[TFW_MMAP_LOG_ADDR]		= {"address", clickhouse::Type::IPv6},
+	[TFW_MMAP_LOG_METHOD]		= {"method", clickhouse::Type::UInt8},
+	[TFW_MMAP_LOG_VERSION]		= {"version", clickhouse::Type::UInt8},
+	[TFW_MMAP_LOG_STATUS]		= {"status", clickhouse::Type::UInt16},
+	[TFW_MMAP_LOG_RESP_CONT_LEN]	= {"response_content_length", clickhouse::Type::UInt32},
+	[TFW_MMAP_LOG_RESP_TIME]	= {"response_time", clickhouse::Type::UInt32},
+	[TFW_MMAP_LOG_VHOST]		= {"vhost", clickhouse::Type::String},
+	[TFW_MMAP_LOG_URI]		= {"uri", clickhouse::Type::String},
+	[TFW_MMAP_LOG_REFERER]		= {"referer", clickhouse::Type::String},
+	[TFW_MMAP_LOG_USER_AGENT]	= {"user_agent", clickhouse::Type::String},
+	[TFW_MMAP_LOG_DROPPED]		= {"dropped_events", clickhouse::Type::UInt64},
+};
+
+std::atomic<bool> stop_flag{false};
+
+#ifdef DEBUG
+static void
+dbg_hexdump(const char *data, int buflen)
+{
+	const unsigned char *buf = (const unsigned char*)data;
+	std::ostringstream oss;
+	oss << std::hex << std::setfill('0');
+
+#define PRINT_CHAR(c) (std::isprint(c) ? c : '.')
+	for (int i = 0; i < buflen; i += 16) {
+		oss << std::setw(6) << i << ": ";
+
+		for (int j = 0; j < 16; ++j)
+			if (i + j < buflen)
+				oss << std::setw(2) << (unsigned)buf[i + j] << " ";
+			else
+				oss << "   ";
+		oss << " ";
+		for (int j = 0; j < 16; ++j) {
+			if (i + j >= buflen)
+				break;
+			oss << (char)PRINT_CHAR(buf[i + j]);
+		}
+		oss << std::endl;
+	}
+	oss << std::dec << "len = " << buflen << std::endl;
+	std::cout << oss.str();
+#undef PRINT_CHAR
+}
+#else
+static void
+dbg_hexdump([[maybe_unused]] const char *data, [[maybe_unused]] int buflen)
+{
+}
+#endif /* DEBUG */
+
+static std::shared_ptr<clickhouse::Block>
+make_block()
+{
+	unsigned int i;
+	auto block = std::make_shared<clickhouse::Block>();
```

Comment:
Are you sure we need a shared_ptr to hold a block? Looks like just a value of `clickhouse::Block` is enough

Reply:
Fixed.